[충남대 FE 김은수] 3단계 로그인 페이지 UI 만들기

shushuburger opened the pull request:
멘토님 안녕하세요. 충남대학교 FE12팀 김은수입니다.

## 과제 요구사항

- [x] 본인만의 기준으로 일관된 코드를 작성함
- [x] 기능 단위로 커밋을 분리함

### 로그인 기능
- [x] `/login` 페이지 접속 시 로그인 화면으로 연결되도록 구현함
- [x] 로그인 화면 UI 구현 완료
- [x] 로그인 버튼 클릭 시 이전 페이지로 Redirect 처리 (없을 경우 `/`으로 이동)
- [x] 네비게이션 바에 로그인 버튼 추가 및 기능 구현
- [x] 네비게이션 바에 뒤로가기 버튼 추가 및 기능 구현

### Not Found 페이지
- [x] 잘못된 경로로 접속 시 Not Found 페이지로 이동되도록 구현함

> 다음의 내용을 src/pages/LoginPage.tsx와 src/pages/NotFoundPage.tsx에 작업했습니다.

### 실시간 급상승 선물랭킹
- [x] 필터(성별, 주제) 선택 시 URL의 query parameter 반영
- [x] 새로고침 후에도 선택한 필터가 유지되도록 구현

> 다음의 내용을 src/components/RankingSection/Ranking*.tsx 에 해당하는 파일들에 작업했습니다.

### 리팩터링
- [x] 재사용 가능성이 높은 컴포넌트를 추출하여 별도 컴포넌트로 분리

> 다음은 리팩터링 대상이 된 파일들 입니다. (리팩터링 전 파일 이름 -> 리팩터링 후 이름)
> 리팩터링 요소들은 모두 src/common에 위치하고 있습니다.

- **RankingSort.tsx → SegmentedControl.tsx**  
  - 기존: 세 개의 고정 정렬 항목을 직접 렌더링
  - 개선: `options`, `selected`, `onSelect` prop을 통해 외부 제어 가능  
  - 사유: 다양한 정렬이나 카테고리 탭 UI에 활용 가능한 컨트롤로 만들기 위해 분리

- **ProductGrid.tsx → CardGrid.tsx**  
  - 기존: 내부에서 3열 고정, `ProductCard`에 종속된 구조  
  - 개선: `columns`, `children` prop으로 다양한 카드형 콘텐츠를 수용 가능  
  - 사유: 2~4열 같이 다양한 레이아웃을 적용할 수 있도록 추상화

- **BannerSection.tsx → TextBanner.tsx**  
  - 기존: 텍스트 내용과 배경색이 하드코딩되어 있었음
  - 개선: `topText`, `bottomText`, `backgroundColor`, `spacingBottom` 등 prop  
  - 사유: 동일한 형태이나 다른 텍스트를 담는 등 다양한 배너를 구현할 때 재사용하기 위해 분리

- **FriendSelectSection.tsx → IconTextButton.tsx**  
  - 기존: 친구 선택 버튼 전용 UI로 구성되어 다른 영역에 재활용 어려움  
  - 개선: `icon`, `text`, `backgroundColor` 등의 prop
  - 사유: 다른 곳에서도 동일한 아이콘+텍스트 버튼 구조 쉽게 이용 가능하도록

shushuburger committed:
refactor(component): 코드 구조를 render → style 순서로 정리

- 컴포넌트의 가독성과 일관성을 위해 구조를 재배치하였습니다.
- render 함수가 상단, styled-components는 하단으로 이동하도록 했습니다.

shushuburger committed:
docs(readme): 3단계 과제 요구사항 체크리스트 추가

- 로그인 페이지, 라우터 구성, 필터 상태 유지 등 핵심 요구사항을 정리했습니다.

shushuburger committed:
refactor(app): 화면 간 이동 위한 기본 구조 생성

- 기존 메인 화면을 pages/HomePage로 분리하였습니다.
- /login, 404에 맞는 페이지 파일 및 라우팅 설정하였습니다.

shushuburger committed:
feat(login): 로그인 페이지 UI 구현 및 리디렉션 처리

- /login 경로에 로그인 컴포넌트를 생성했습니다.
- 이메일, 비밀번호 입력 영역과 로그인 버튼으로 UI를 구성했습니다.
- 로그인 버튼 클릭하면 location.state?.from 기반 이전 페이지로 이동이 이루어지게 했습니다.
- 이전 페이지 정보 없을 경우 /로 이동하도록 설정해두었습니다.
- README에 구현 완료를 표시했습니다.

shushuburger committed:
feat(navigation): 로그인/뒤로가기/로고 버튼 기능 추가

- 로그인 버튼 클릭 시 /login으로 이동하도록 했습니다.
- 로그인 화면(/login)에서는 로그인 버튼 비활성화되도록 했습니다.
- 뒤로가기 버튼은 navigate(-1)로 작동하도록 구현하였습니다.
- 로고 클릭 시 항상 메인(/)으로 이동하도록 작성해두었습니다.
- README 구현 완료 현황을 반영하였습니다.

shushuburger committed:
feat(ranking): 쿼리 기반 정렬 필터 적용 및 새로고침 시 상태 유지 기능 추가

- RankingSort에 label-value 구조를 이용하였습니다.
- 화면에는 한글 라벨 표시가 되고 쿼리에는 영문으로 참고 링크와 동일하게 뜨도록 하였습니다.
- 선택된 필터의 값이 URL에 쿼리 파라미터로 반영되도록 구현하였습니다.
- 새로고침 시에도 쿼리 파라미터를 기반으로 URL과 버튼 선택 상태를 유지하도록 했습니다.

shushuburger committed:
refactor(ranking): 불필요한 변수 제거

- 사용되지 않는 코드를 삭제했습니다.

shushuburger committed:
fix(auth): 로그인 후 쿼리 파라미터까지 포함해 이전 페이지로 정확히 리디렉션되도록 수정

- LoginPage.tsx에서 location.state.from?.pathname + search 조합으로 redirect 경로를 구성했습니다.
- Navigation.tsx에서 로그인 버튼 클릭 시 현재 location을 state로 전달해 from 정보를 유지하도록 했습니다.
- 이전 페이지가 없을 경우 기본값으로 '/'로 이동하도록 작성해주었습니다.

shushuburger committed:
refactor(components): 재사용 가능한 UI 컴포넌트 분리

- 재사용성이 높다고 판단한 컴포넌트를 분리하였습니다.
- README에 구현 완료를 표시했습니다.

shushuburger committed:
style(components): 컴포넌트 내 코드 구조를 interface → render → style 순으로 정렬

shushuburger committed:
refactor(import): 컴포넌트 import 경로 절대경로로 통일

shushuburger committed:
feat(not-found): NotFoundPage UI 구현 및 배경색 회색 처리

- Navigation 포함, 이미지/텍스트/버튼 구성 UI 구현하였습니다.
- '홈으로' 버튼 클릭 시 메인 페이지('/')로 이동 처리하였습니다.

igy95 commented on src/App.tsx:
```diff
@@ -1,19 +1,20 @@
+        <Routes>
+          <Route path="/" element={<HomePage />} />
+          <Route path="/login" element={<LoginPage />} />
+          <Route path="*" element={<NotFoundPage />} />
```
Route 경로를 상수로 관리해보면 어떨까요? 단순 값으로 처리하는 것보다 의미 전달이나 재사용이 더 용이할 수 있을 것 같습니다~

shushuburger replied:
src/constants/routes.ts 파일을 생성하여 아래와 같이 상수 처리를 해주었습니다.

```
export const ROUTES = {
  HOME: '/',
  LOGIN: '/login',
  NOT_FOUND: '*',
} as const;
```

이후 App.tsx와 Navigation.tsx에서 ROUTES를 import하여 상수를 사용하도록 해주었습니다!

igy95 commented on src/components/common/TextBanner.tsx:
Banner에 텍스트가 아닌 다른 요소가 들어오는 경우에도 재사용이 가능하도록 재사용성을 높여보면 어떨까요?

shushuburger replied:
기존에는 무조건 string 2개만을 받던 구조를 children을 받도록 하였습니다.
파일명과 관련된 변수 모두 TextBanner -> Banner로 변경하여 조금 더 일반적인 의미를 지니도록 했습니다.

https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/basic_type_example
https://react.dev/learn/passing-props-to-a-component#passing-jsx-as-children

두 링크를 참고하여 children에 대해 지식을 습득했습니다.

igy95 commented on src/components/Navigation.tsx:
```diff
@@ -1,37 +1,16 @@
 const Navigation = () => {
+  const location = useLocation();
+  const navigate = useNavigate();
+  const isLoginPage = location.pathname === '/login';
```
route path를 상수로 관리하면 여기서 재사용이 가능해보이네요 😀

shushuburger replied:
제일 첫번째 피드백을 반영하며 해당 코드 역시 다음과 같이 상수를 이용하는 방식으로 변경해주었습니다.

`const isLoginPage = location.pathname === ROUTES.LOGIN;`

igy95 commented on src/components/FriendSelectSection.tsx:
```diff
@@ -1,48 +1,18 @@
-      <SelectButton aria-label="선물할 친구 선택">
-        <IconWrapper>
+    <section
+      style={{
+        boxSizing: 'border-box',
+        height: '106px',
+        padding: '16px 12px',
+        backgroundColor: '#f2f3f6',
+        marginBottom: '16px',
+      }}
```
개인적인 의견일 수 있지만, 특별한 의도가 있지 않다면 스타일 방식은 되도록 하나의 방식으로 통일해서 사용하는 게 컴포넌트에 대한 가독성을 높이는 방식이라는 생각이 듭니다! 이것도 스타일 라이브러리로 따로 빼면 어떨까요?

shushuburger replied:
```
const Section = styled.section`
  box-sizing: border-box;
  height: 106px;
  padding: ${({ theme }) => `${theme.spacing[4]} ${theme.spacing[3]}`};
  background-color: ${({ theme }) => theme.color.gray[200]};
  margin-bottom: ${({ theme }) => theme.spacing[4]};
`;
```

다른 파일들과 동일한 형태로 작성해주며 theme을 사용해 값을 지정해주었습니다!

igy95 commented on src/components/Navigation.tsx:
```diff
@@ -1,37 +1,16 @@
     <Nav>
       <div>
-        <IconButton aria-label="뒤로가기">
+        <IconButton onClick={() => navigate(-1)} aria-label="뒤로가기">
```
이렇게 처리하면, 앱 외에 다른 화면에서 앱 진입 후 바로 눌렀을 때 앱 안에 머무르지 않고 다시 앱 바깥의 이전 화면으로 돌아가게 되지 않을까요?

shushuburger replied:
네 그렇습니다.

과제에 대해 제공된 참고 링크를 눌러보았을 때도 "naver -> 앱" 과 같은 경로로 이동한 다음 "뒤로 가기"를 누르면
naver로 이동하는 것을 확인했었습니다.
뒤로가기의 범위에 대해 따로 언급이 없었기에 참고 링크와 최대한 동작을 같게 하려하였습니다.
그래서 현재 제 코드도 "chrome -> 앱"과 같은 상태에서 바로 뒤로 가기를 누르면 "chrome"으로 이동합니다.

만약 "뒤로 가기"를 눌렀을 때 앱 안에만 머무를 수 있도록 수정해야 한다면 이 부분은 조금 더 방법을 연구한 다음에 2단계 과제를 진행하며 반영해도 괜찮을지 여쭈어봅니다.

igy95 replied:
아 꼭 수정해야한다, 는 아니었고 이러한 UX 처리에서 엣지 케이스가 될 수 있는 하나의 예시를 공유 드리고자한 거였어요 !! 참고 링크의 동작과 동일하게 가져가는 의도도 이해가 되기 때문에 나중에 고민해보셔도 좋겠습니다 ~

igy95 commented on src/components/Navigation.tsx:
```diff
@@ -49,11 +28,23 @@ const Navigation = () => {
+            })
+          }
+          aria-label="로그인"
+          disabled={isLoginPage}
```
현재의 구현 방향은, Navigation 안에서 각 페이지의 제약 사항을 알고 있는 구조입니다. 이러한 방식은 페이지가 점점 많아질수록 Navigation 안에서 알아야할 사항이 늘어남을 의미할 수 있을 것 같아요. Navigation에서 페이지를 아는 것이 아닌, 페이지에서 Navigation을 알도록 처리하면 어떨까요?

shushuburger replied:
```
interface NavigationProps {
  showLoginButton?: boolean;
}
```

위와 같이 로그인 화면인지 아닌지를 알 수 있는 boolean 변수를 만들어준 다음 LoginPage에서 Navigation 부분에 props로 해당 변수의 값을 주도록 했습니다.
이 방법을 통해 Navigation.tsx는 UI의 기능에만 집중할 수 있고 각각의 페이지가 어떤 페이지인지 몰라도 알맞은 UI를 보일 수 있도록 했습니다.

igy95 replied:
현재의 수정 방향도 좋습니다. 여기서 더 유연하게 처리하고자 한다면 아래와 같이 Navigation이 컴포넌트를 받도록 slot을 열어두는 것도 괜찮을 것 같아요!

```ts
<Navigation rightSection={<LoginButton />} />
```

igy95 commented on src/components/RankingSection/RankingFilter.tsx:
```diff
@@ -5,6 +5,34 @@ interface RankingFilterProps {
 }
 
+const RankingFilter = ({ selectedFilter, onSelect }: RankingFilterProps) => {
+  const filters = [
+    { emoji: 'ALL', label: '전체', value: 'ALL' },
+    { emoji: '👩🏻', label: '여성이', value: 'FEMALE' },
+    { emoji: '👨🏻', label: '남성이', value: 'MALE' },
+    { emoji: '👦🏻', label: '청소년이', value: 'TEEN' },
+  ];
```
렌더링마다 재선언 되지 않도록 컴포넌트 바깥으로 빼는 게 좋을 듯 합니다!

shushuburger replied:
컴포넌트 코드 바깥으로 이동시켰습니다!

```
const filters = [
  { emoji: 'ALL', label: '전체', value: 'ALL' },
  { emoji: '👩🏻', label: '여성이', value: 'FEMALE' },
  { emoji: '👨🏻', label: '남성이', value: 'MALE' },
  { emoji: '👦🏻', label: '청소년이', value: 'TEEN' },
];

const RankingFilter = ({ selectedFilter, onSelect }: RankingFilterProps) => {
```

RankingSort.tsx에서도 동일한 구조를 발견하여 이 역시 컴포넌트 바깥으로 코드 위치를 옮겼습니다!
```
const options = [
  { label: '받고 싶어한', value: 'MANY_WISH' },
  { label: '많이 선물한', value: 'MANY_RECEIVE' },
  { label: '위시로 받은', value: 'MANY_WISH_RECEIVE' },
];

const RankingSort = ({ selectedSort, onSelect }: RankingSortProps) => {
  return (
    <SegmentedControl
      options={options}
      selectedValue={selectedSort}
      onSelect={onSelect}
    />
  );
};
```

igy95 commented on src/components/common/IconTextButton.tsx:
IconTextButton을 따로 만들기보다 아예 Button이라는 컴포넌트를 만들어서 그 안에서 처리를 할 수 있다면 재사용성이 더 늘어날 듯합니다!

shushuburger replied:
기존 IconTextButton의 아래와 같은 한계를 극복하기 위해 Button.tsx를 새로 만들어 작업했습니다!
<개선 전>
- 아이콘 반드시 있어야 함
- 텍스트로 이루어져야 함

<개선 후>
- 아이콘 유무 선택 가능
- 텍스트 외에 다양한 요소가 들어갈 수 있음

igy95 commented on src/pages/LoginPage.tsx:
```diff
@@ -0,0 +1,81 @@
+const LoginPage = () => {
+  const navigate = useNavigate();
+  const location = useLocation();
+  const from = location.state?.from;
```
route로 받는 상태의 단점은, 파일이 많아지는 경우 이 상태를 어디로부터 받아오는지 파악이 어렵다는 데 있습니다. 때문에 이러한 부분을 코드로 보완해주기 위해서는 `from`과 같은 key를 상수로 따로 두고 연관되는 부분을 맵핑 시키는 방식도 고려해볼 수 있을 것 같아요!

shushuburger replied:
- location.state?.from처럼 문자열 key를 하드코딩하면 추적이 어려움
- 키 이름('from')에 오타가 나도 컴파일러는 감지하지 못함
- 프로젝트가 커지면 전달자와 수신자 사이의 연결을 파악하기 어려움

gpt를 활용하여 from과 같은 key에 대해 더 알아본 결과 위와 같은 단점이 있다는 것을 추가적으로 알았습니다!

src/constants/navigationState.ts를 생성하고 아래와 같이 코드를 작성하여 상수화 시켜주었습니다!
```
export const LOCATION_STATE_KEYS = {
  FROM: 'from',
} as const;
```

igy95 reviewed (changes requested):
안녕하세요 은수님~ 페이지 UI 구현 외에도 재사용성을 높이기 위한 시도를 해보신 부분 좋습니다! 코멘트 확인 후 재요청 부탁드려요 :)

shushuburger committed:
refactor(routing): route 경로를 상수로 분리하고 App 및 Navigation에서 재사용

- router.ts 파일에 ROUTES 상수를 정의했습니다. (HOME, LOGIN, NOT_FOUND)
- App.tsx의 Route path를 ROUTES 상수로 치환하였습니다.
- Navigation.tsx에서도 location.pathname === ROUTES.LOGIN 형태로 변경해주었습니다.

shushuburger committed:
refactor(friend-select): 스타일을 Emotion styled 방식으로 통일

- styled-components 방식으로 스타일을 정의해 통일감을 높였습니다.

shushuburger committed:
refactor(ranking): 컴포넌트의 코드 구성 순서 정리

- RankingFilter.tsx에서 filters 배열을 컴포넌트 외부로 이동해 불필요한 재선언을 방지했습니다.
- RankingSort.tsx에서 정렬 옵션도 외부 상수로 선언하고 SegmentedControl로 일관된 UI를 적용했습니다.

shushuburger committed:
refactor(login): location.state key를 navigationState.ts로 분리

- 'from' 문자열 key를 LOCATION_STATE_KEYS.FROM 상수로 정의하여 사용했습니다.

shushuburger committed:
refactor(text-banner): children 기반 구조로 개선

- TextBanner 컴포넌트가 children을 받아 다양한 JSX 콘텐츠를 지원하도록 개선하였습니다.
- 기존의 topText, bottomText 문자열 props를 제거하였습니다.
- BannerSection.tsx도 이에 맞춰 <Banner>...</Banner> 형태로 수정하였습니다.

shushuburger committed:
refactor(friend-select): IconTextButton를 Button 컴포넌트로 교체

- FriendSelectSection.tsx에서 IconTextButton 제거 후 Button 컴포넌트로 변경하였습니다.

shushuburger committed:
refactor(navigation): 페이지 판단을 props로 책임 분리

- isLoginPage 조건을 제거했습니다.
- showLoginButton props를 Navigation이 전달받아 처리하도록 변경하였습니다.
- 페이지별로 Navigation 동작을 제어하도록 구조를 개선하였습니다.

igy95 reviewed (approved):
안녕하세요 은수님 ~ 본인이 이해한 바를 토대로 리뷰를 어떻게 수정했는지, 어떤 부분을 참고했는지 의도를 잘 작성해주셔서 리뷰하기가 수월했습니다 ! 다음 단계도 화이팅하시길 바라요 👍
